simtools/model/telescope_model.py

```diff
@@ -763,3 +779,81 @@ def getOnAxisEffOpticalArea(self):
             )
             raise ValueError
         return rayTracingData["eff_area"][0]
+
+    def readIncidenceAngleDistribution(self, incidenceAngleDistFile):
+        """
+        Read the incidence angle distrubution from a file
+
+        Parameters
+        ----------
+        incidenceAngleDistFile: str
+            File name of the incidence angle distribution
+
+        Returns
+        -------
+        incidenceAngleDist: Astropy table
+            Astropy table with the incidence angle distribution
+        """
+
+        incidenceAngleDist = astropy.io.ascii.read(
+            self.getDerivedDirectory().joinpath(incidenceAngleDistFile)
+        )
+        return incidenceAngleDist
+
+    @staticmethod
+    def calcAverageCurve(curves, incidenceAngleDist):
+        """
+        Calculate an average curve from a set of cuvres, using as weights
+        the distribution of incidence angles provided in incidenceAngleDist
+
+        Parameters
+        ----------
+        curves: dict
+            dict of with 3 "columns", Wavelength, Angle and z
+            The dictionary represents a two dimensional distribution of wavelengths and angles
+            with the z value being e.g., reflecitivity, transmission, etc.
+        incidenceAngleDist: Astropy table
+            Astropy table with the incidence angle distribution
+            The assumed columns are "Incidence angle" and "Fraction".
+
+        Returns
+        -------
+        averageCurve: Astropy Table
+            Table with the averaged curve
+        """
+
+        weights = list()
+        for angleNow in curves["Angle"]:
+            weights.append(
+                incidenceAngleDist["Fraction"][
```

Comment:
In this step we calculate the weight of each curve in the two-dimensional wavelength/angle/z curve. The incidence angle table looks like the following

```
"Incidence angle" Fraction
0.1 0.0
0.3 0.0
0.5 0.0
0.7 0.0
0.9 0.0
1.1 0.0
1.3 0.0
1.5 0.0
1.7 0.0
1.9 0.0
2.1 0.0
2.3 0.0
2.5 0.0
2.7 0.0
2.9 0.0
...
23.1 0.0
23.3 0.0
23.5 1.2284690327526267e-05
23.7 0.00013922649037863104
23.9 0.00019655504524042027
24.1 0.00036444581304993947
24.3 0.0004176794711358931
24.5 0.0006756579680139447
24.7 0.0009991548133054698
24.9 0.001285797587614416
25.1 0.0014291189747688638
25.3 0.001887747413663203
25.5 0.0022521932267131487
25.7 0.002477412549384464
25.9 0.0024118942009709906
26.1 0.0027149165623832567
26.3 0.0031940194851568294
26.5 0.0037836846208780905
26.7 0.004107181466169616
26.9 0.004750080259976823
27.1 0.005233278079526097
27.3 0.0055608698215935575
27.5 0.006596878705881605
27.7 0.006977704106034921
27.9 0.007698405938583127
28.1 0.008529669984078921
...
59.3 0.014450890721946476
59.5 0.0077925885644276344
59.7 0.00033168663884320336
59.9 0.0
60.1 0.0
60.3 0.0
60.5 0.0
60.7 0.0
...
```

Then the 2D curve we are trying to average looks like 

```
ANGLE=    0             10             20           30           40           50           60             70
#
200	0.238501       0.0283268      0.0533797    0.0504117    0.0474436    0.0439879    0.0405321      0.0401532
205	0.188644       0.0175978      0.0475514    0.0387747    0.0299981    0.0302341    0.0304701      0.0213979
210	0.179222       0.0281071      0.0484482    0.0338117    0.0191752    0.0217524    0.0243297      0.00887241
...
```

So, for each angle in the 2D curve I search for the closest entry in the incidence angle table and that would be the weight. The majority in this case are zero, but in the SST focal plane case we have more entries in the 2D curve and the average is more meaningful. I could also add an interpolation step that adds curves to the 2D distribution, but I don't think it would make a real difference. The important thing is that this function is general enough to deal with any kind of tables we will get in the future.

Does this make sense? I will wait with the merge until you confirm since this is a basic logic point, not a minor detail.

Reply:
yes, this make sense - I understand the code now. So fine with merging.